src/OpenTelemetry.Instrumentation.Wcf/Implementation/HttpRequestMessagePropertyWrapper.cs

```diff
@@ -0,0 +1,102 @@
+// <copyright file="HttpRequestMessagePropertyWrapper.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Net;
+using System.Reflection;
+
+namespace OpenTelemetry.Instrumentation.Wcf.Implementation;
+
+/// <summary>
+/// This is a reflection-based wrapper around the HttpRequestMessageProperty class. It is done this way so we don't need to
+/// have an explicit reference to System.ServiceModel.Http.dll. If the consuming application has a reference to
+/// System.ServiceModel.Http.dll then the HttpRequestMessageProperty class will be available (IsHttpFunctionalityEnabled == true).
+/// If the consuming application does not have a reference to System.ServiceModel.Http.dll then all http-related functionality
+/// will be disabled (IsHttpFunctionalityEnabled == false).
+/// </summary>
+internal class HttpRequestMessagePropertyWrapper
+{
+    private static readonly ReflectedInfo ReflectedValues = Initialize();
+    private readonly object innerObject;
+
+    public HttpRequestMessagePropertyWrapper(object existingHttpRequestMessageProperty = null)
+    {
+        EnsureEnabled();
+        if (existingHttpRequestMessageProperty != null && !existingHttpRequestMessageProperty.GetType().Equals(ReflectedValues.Type))
+        {
+            throw new ArgumentException("Existing object must be of type HttpRequestMessageProperty", nameof(existingHttpRequestMessageProperty));
+        }
+
+        this.innerObject = existingHttpRequestMessageProperty != null ? existingHttpRequestMessageProperty : Activator.CreateInstance(ReflectedValues.Type);
+    }
+
+    public static bool IsHttpFunctionalityEnabled => ReflectedValues != null;
+
+    public static string Name
+    {
+        get
+        {
+            EnsureEnabled();
+            return (string)ReflectedValues.NameProperty.GetValue(null);
+        }
+    }
+
+    public WebHeaderCollection Headers => (WebHeaderCollection)ReflectedValues.HeadersProperty.GetValue(this.innerObject);
+
+    public object InnerObject => this.innerObject;
+
+    private static ReflectedInfo Initialize()
```

Comment:
@repl-chris This is nice thanks for doing it! A little bit of feedback...

* `Name` seems to be static so we could probably just cache the result?
* `Headers` we are invoking through the `MethodInfo` which is pretty slow. It is better to create a delegate from that. We have this class `PropertyFetcher` you can link into the project which will help with that. I have some benchmarks below that show the improvement. Doing this exercise I noticed the perf of `PropertyFetcher` is kind of crappy. On the benchmarks I have a lower-level version which is even better. I think though, use `PropertyFetcher` and we can improve its performance which will then benefit everything else?

|                              Method |      Mean |     Error |    StdDev | Allocated |
|------------------------------------ |----------:|----------:|----------:|----------:|
|      RetrieveHeadersUsingReflection | 56.310 ns | 0.8513 ns | 0.7546 ns |         - |
|        RetrieveHeadersUsingDelegate |  6.571 ns | 0.1724 ns | 0.1917 ns |         - |
| RetrieveHeadersUsingPropertyFetcher | 16.139 ns | 0.2702 ns | 0.2527 ns |         - |

```csharp
[MemoryDiagnoser]
public class WcfReflectionBenchmarks
{
    private static readonly ReflectedInfo ReflectedValues = Initialize();

    private readonly HttpRequestMessageProperty request = new();

    [Benchmark]
    public WebHeaderCollection RetrieveHeadersUsingReflection()
    {
        return (WebHeaderCollection)ReflectedValues.HeadersProperty.GetMethod.Invoke(this.request, null);
    }

    [Benchmark]
    public WebHeaderCollection RetrieveHeadersUsingDelegate()
    {
        return ReflectedValues.GetHeaders.Invoke(this.request);
    }

    [Benchmark]
    public WebHeaderCollection RetrieveHeadersUsingPropertyFetcher()
    {
        return ReflectedValues.GetHeadersPropertyFetcher.Fetch(this.request);
    }

    private static ReflectedInfo Initialize()
    {
        var type = Type.GetType(
            "System.ServiceModel.Channels.HttpRequestMessageProperty, System.ServiceModel, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089",
            true);
        var headersProp = type.GetProperty("Headers", BindingFlags.Public | BindingFlags.Instance, null, typeof(WebHeaderCollection), Array.Empty<Type>(), null);
        var nameProp = type.GetProperty("Name", BindingFlags.Public | BindingFlags.Static, null, typeof(string), Array.Empty<Type>(), null);

        if (type != null && headersProp != null && nameProp != null)
        {
            return new ReflectedInfo
            {
                Type = type,
                Name = (string)nameProp.GetMethod.Invoke(null, null),
                HeadersProperty = headersProp,
                GetHeaders = ReflectedInfo.CreateDelegateWrapper<WebHeaderCollection>(headersProp.GetMethod),
                GetHeadersPropertyFetcher = new PropertyFetcher<WebHeaderCollection>("Headers"),
            };
        }

        throw new NotSupportedException("WCF reflection failed to bind");
    }

    private sealed class ReflectedInfo
    {
        public Type Type;
        public string Name;
        public PropertyInfo HeadersProperty;
        public DelegateWrapper<WebHeaderCollection> GetHeaders;
        public PropertyFetcher<WebHeaderCollection> GetHeadersPropertyFetcher;

        public static DelegateWrapper<T> CreateDelegateWrapper<T>(MethodInfo getMethodInfo)
        {
            var delegateWrapperStronglyTyped = Activator.CreateInstance(
                typeof(DelegateWrapperStronglyTyped<,>).MakeGenericType(getMethodInfo.DeclaringType, getMethodInfo.ReturnType),
                BindingFlags.Public | BindingFlags.Instance,
                binder: null,
                args: new object[] { getMethodInfo },
                culture: null);

            return (DelegateWrapper<T>)delegateWrapperStronglyTyped;
        }

        public abstract class DelegateWrapper<T>
        {
            public abstract T Invoke(object instance);
        }

        private sealed class DelegateWrapperStronglyTyped<TClass, TProperty> : DelegateWrapper<TProperty>
        {
            private readonly Func<TClass, TProperty> innerDelegate;

            public DelegateWrapperStronglyTyped(MethodInfo getMethodInfo)
            {
                this.innerDelegate = (Func<TClass, TProperty>)getMethodInfo.CreateDelegate(typeof(Func<TClass, TProperty>));
            }

            public override TProperty Invoke(object instance)
                => this.innerDelegate((TClass)instance);
        }
    }
}
```

Reply:
Worth to check if we can just grab newer version PropertyFetcher from main repository. I suppose it was not synced for a while.

Reply:
I have updated the reflection as suggested.  I looked into grabbing the newer version of PropertyFetcher, but it has breaking changes so I backed away slowly...probably should do that in a separate PR.  

On a side note, anyone have any tips/tricks on how to easily test those EventSource logs?...I was able to manually test the new one, but it took me way too long to figure out and I'm sure there must be an easier way than what I did 😀

Reply:
I have created #1311 to track this update. It will be great if you would like to contribute. For sure, it is not a blocker for this PR.

Reply:
@repl-chris RE: The EventSource what you want to do? Set up a listener? This class might help you with that: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/test/OpenTelemetry.Contrib.Tests.Shared/TestEventListener.cs

IIRC @TimothyMothra has some experience testing EventSources. Depending on what you want to do he may be able to help as well.

Reply:
Oooooh yeah, that's exactly what I didn't know I was looking for...I've never played with ETW events before so I was running around in the dark bumping into things 😂

Reply:
@repl-chris Hey sorry I haven't approved/merged this because it sounded like you were going to do some kind of test for the EventSource. I don't think it is super important that we do that, but if you want to by all means feel free. Just LMK if you are all done.